### PR TITLE
AB#38235 merge switching perms

### PR DESCRIPTION
**Pull request by lhao03**
## Overview

* we need a button to switch between user mode and Teams app mode. Right now this button:
* changes the header from "Graph Explorer (as user)" to  "Graph Explorer (as Teams app)".
* changes the blurb in the "Modify Permissions (Preview)" permissions tab from "One of the following delegated permissions is required to run the query. Select a permission and click Consent." from "One of the following resource specific consent permissions is required to run the query. Select a permission and click Consent.".
* changes the permissions from DelegatedWork to Application in the "Modify Permissions (Preview)" tab.
* right now can't accept Application permissions
* accepting delegated permissions still works.
* I added RSC permissions filtering and merged the other branch in this PR #9 

### Demo

perms
perms-filtering

### Notes

* ~~right now someone can't use Teams app mode when not signed in. Do we want to keep it that way?~~ resolved.

## Testing Instructions

* Start after checking out this branch
* If you're using WSL, run `nvm use 14.0`.
* then run `npm install && npm start`
* everything should work as in the gif.
* I added tests for the reducer and action creator in files src/tests/services/actions/permission-mode-action-creators.spec.ts and src/tests/services/reducers/permission-mode-reducers.spec.tsx

### CODE REVIEW

File: src/telemetry/filters.ts [hunk 1]
```diff
@@ -1,5 +1,4 @@
 import { ITelemetryItem } from '@microsoft/applicationinsights-web';
-import { exception } from 'console';
```
Comment (omersayshi) [inline src/telemetry/filters.ts:2]: Why was this removed? It doesnt seem to be related to the task or anything in the PR

Comment (lhao03) [reply src/telemetry/filters.ts:2]: It was causing a linting error so I removed it

Comment (omersayshi) [reply src/telemetry/filters.ts:2]: Sounds good!

**Comment by omersayshi** [comment]
Initially scanning the changes, things look good. @lhao03 I do not see any test additions or changes, so you probably need to do that

File: src/app/views/app-sections/AppTitle.tsx [hunk 2]
```diff
@@ -49,16 +62,33 @@ export function appTitleDisplayOnMobileScreen(
...
   )
 }
+
+function permissionsModeButton(changeMode: Function, permissionType: boolean) {
+  return (
+    <IconButton
+      iconProps={{ iconName: 'Cat' }}
+      onClick={() => changeMode(!permissionType)}
+      ariaLabel="button to switch between user and Teams app" />
+  )
```
Comment (lhao03) [inline src/app/views/app-sections/AppTitle.tsx:93]: This will need to change.

File: src/app/views/app-sections/AppTitle.tsx [hunk 3]
```diff
@@ -49,16 +62,33 @@ export function appTitleDisplayOnMobileScreen(
...
   return (
     <Label className={classes.graphExplorerLabel}>
-      Graph Explorer
+      Graph Explorer {permissionType ? "(as user)" : "(as Teams app)"}
     </Label>
   )
 }
+
+function permissionsModeButton(changeMode: Function, permissionType: boolean) {
+  return (
```
Comment (lhao03) [inline src/app/views/app-sections/AppTitle.tsx:88]: This button will need to change but here for now for testing purposes.

File: src/app/views/app-sections/AppTitle.tsx (hunk 2 above)
Comment (lhao03) [inline src/app/views/app-sections/AppTitle.tsx:93]: This button is for testing purposes, and is not the final form.

File: src/app/views/authentication/profile/Profile.tsx [hunk 4]
```diff
@@ -157,7 +157,7 @@ export class Profile extends Component<IProfileProps, IProfileState> {
     };
 
     return (
-      <div className={classes.profile}>
+      <div className={classes.profile} style={{ display: 'flex', width: '115%' }}>
```
Comment (lhao03) [inline src/app/views/authentication/profile/Profile.tsx]: I think @gracet9182 added this. I don't recall adding it. What's the purpose? It doesn't seem to break anything.

File: src/app/views/query-runner/request/permissions/TabList.tsx [hunk 5]
```diff
@@ -18,11 +18,21 @@ interface ITabList {
...
+    return RSC_PERMISSIONS_ENDINGS.some((ending) => permission.value.indexOf(ending) !== -1)
+  }
+
+  const filterPermissionsForRSC = () => {
+    return permissionModeType === DISPLAY_APPLICATION_PERMISSIONS ? permissions.filter(isRSC) : permissions;
+  }
+
+  const filteredPermissions = filterPermissionsForRSC();
+
+  setConsentedStatus(tokenPresent, filteredPermissions, consentedScopes);
```
Comment (lhao03) [inline src/app/views/query-runner/request/permissions/TabList.tsx:35]: Might need to double check this.

File: src/app/services/graph-constants.ts [hunk 6]
```diff
@@ -1,8 +1,15 @@
...
+export const USER_BETA_INFO_URL = `${GRAPH_URL}/beta/me/profile`;
 export const USER_PICTURE_URL = `${GRAPH_URL}/beta/me/photo/$value`;
 export const AUTH_URL = 'https://login.microsoftonline.com';
 export const DEFAULT_USER_SCOPES = 'openid profile User.Read';
 export const DEVX_API_URL = 'https://graphexplorerapi.azurewebsites.net';
 export const GRAPH_API_SANDBOX_URL = 'https://proxy.apisandbox.msdn.microsoft.com';
 export const HOME_ACCOUNT_KEY = 'fbf1ecbe-27ab-42d7-96d4-3e6b03682ee4';
+export const APPLICATION = "Application";
+export const WORK = "DelegatedWork";
+export const PERSONAL = "DelegatedPersonal";
```
Comment (acchiang) [inline src/app/services/graph-constants.ts]: could these constant names be more descriptive? we're talking about perm scope modes here right, what do you think about PERM_SCOPE_APPLICATION, PERMS_SCOPE_WORK?

File: src/app/services/graph-constants.ts [hunk 7]
```diff
@@ -1,8 +1,15 @@
 export const GRAPH_URL = 'https://graph.microsoft.com';
 export const USER_INFO_URL = `${GRAPH_URL}/v1.0/me`;
+export const USER_BETA_INFO_URL = `${GRAPH_URL}/beta/me/profile`;
```
Comment (acchiang) [inline src/app/services/graph-constants.ts]: I don't see this const being used anywhere else in your PR, was this added intentionally?

File: src/app/services/graph-constants.ts (hunk 6 above)
Comment (acchiang) [reply src/app/services/graph-constants.ts]: i also noticed PERSONAL isn't used in this PR (and delegated personal isn't supported by MS Graph iirc), would you agree that we should avoid adding unused code even if it's a constant that matches the rest of the set?

File: src/app/services/reducers/permission-mode-reducer.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+} from '../graph-constants';
+
+export function permissionModeType(state: boolean = DISPLAY_DELEGATED_PERMISSIONS, action: IAction): any {
+    switch (action.type) {
+        case CHANGE_PERMISSIONS_MODE_SUCCESS:
+            return action.response;
+        default:
+            return state;
+    }
+}
```
Comment (acchiang) [inline src/app/services/reducers/permission-mode-reducer.ts]: not sure if this is something you feel as well but this "no newline" error bugs me all the time :)

File: src/app/views/app-sections/AppTitle.tsx [hunk 9]
```diff
@@ -28,16 +30,27 @@ export function appTitleDisplayOnFullScreen(
     <div className={classes.graphExplorerLabelContainer} role={'heading'} aria-level={1}>
       {!minimised &&
         <>
-          {displayGraphLabel(classes)}
+          {displayGraphLabel(classes, permissionType)}
         </>}
     </div>
+    <div style={{ marginTop: 15 }}>
```
Comment (acchiang) [inline src/app/views/app-sections/AppTitle.tsx]: at previous companies I worked at, there were style guidelines that restricted "magic number" css values (ie. all margin values need to be a multiple of 4px, for example). 
could you take a look at other divs/marginTops in the codebase and make sure that 15px is a value that's used often for margins?

File: src/app/views/app-sections/AppTitle.tsx (hunk 2 above)
Comment (acchiang) [reply src/app/views/app-sections/AppTitle.tsx:93]: image

File: src/app/views/authentication/Authentication.tsx [hunk 10]
```diff
@@ -36,6 +36,7 @@ const Authentication = (props: any) => {
 
     try {
       const authResponse = await authenticationWrapper.logIn();
+
```
Comment (acchiang) [inline src/app/views/authentication/Authentication.tsx]: should we remove this? haha

**Review by acchiang — COMMENTED** [review]
left some comments! haven't been able to make it through the entire PR yet though. would be happy to take a look again :)

File: src/app/services/actions/permission-mode-action-creator.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { IAction } from '../../../types/action';
```
Comment (csyan5) [inline src/app/services/actions/permission-mode-action-creator.ts:1]: Typescript allows to config for a non-absolute way for importing modules if you feel there're too many `../../` 

https://www.typescriptlang.org/tsconfig#baseUrl

File: src/app/services/actions/permissions-action-creator.ts [hunk 12]
```diff
@@ -13,6 +12,12 @@ import {
   FETCH_SCOPES_PENDING,
   FETCH_SCOPES_SUCCESS,
 } from '../redux-constants';
+import {
+  WORK,
```
Comment (csyan5) [inline src/app/services/actions/permissions-action-creator.ts]: These can be captured as Enum 

https://www.typescriptlang.org/docs/handbook/enums.html

File: src/app/services/actions/permissions-action-creator.ts [hunk 13]
```diff
@@ -54,7 +59,15 @@ export function fetchScopes(): Function {
           throw new Error('url is invalid');
         }
 
-        permissionsUrl = `${permissionsUrl}?requesturl=/${requestUrl}&method=${query.selectedVerb}`;
+        let scope = WORK;
+
+        if (permissionModeType === DISPLAY_APPLICATION_PERMISSIONS) {
```
Comment (csyan5) [inline src/app/services/actions/permissions-action-creator.ts]: Can use a map to reduce if statements

File: src/app/views/app-sections/AppTitle.tsx (hunk 9 above)
Comment (csyan5) [inline src/app/views/app-sections/AppTitle.tsx]: Can we separate the CSS in the .scss files?

File: src/app/views/app-sections/AppTitle.tsx (hunk 2 above)
Comment (lhao03) [reply src/app/views/app-sections/AppTitle.tsx:93]: omg I laughed out loud at this 🤣

File: src/app/services/graph-constants.ts (hunk 6 above)
Comment (lhao03) [reply src/app/services/graph-constants.ts]: yeah i just added it for sake of matching the set but it isn't used and might confuse people. i'll remove it

File: src/app/services/graph-constants.ts (hunk 7 above)
Comment (lhao03) [reply src/app/services/graph-constants.ts]: i was trying out random stuff, thanks for catching that LOL. will remove
YikesMonkeyLookAndLeaveGIF

File: src/app/views/app-sections/AppTitle.tsx (hunk 9 above)
Comment (lhao03) [reply src/app/views/app-sections/AppTitle.tsx]: yup there are definitely magic values 15pxx is not it '😅. i think the marginTop was added in for the toggle because the toggle takes up some space so ill remove all the newly added css floating around.

File: src/app/views/authentication/Authentication.tsx (hunk 10 above)
Comment (lhao03) [reply src/app/views/authentication/Authentication.tsx]: the empty line has been yeeted

File: src/app/services/actions/permissions-action-creator.ts (hunk 12 above)
Comment (lhao03) [reply src/app/services/actions/permissions-action-creator.ts]: thanks for the suggestion! using enum makes alot more sense 😊

File: src/app/services/actions/permission-mode-action-creator.ts (hunk 11 above)
Comment (lhao03) [reply src/app/services/actions/permission-mode-action-creator.ts:1]: i think all the action creators have it imported as `import { IAction } from '../../../types/action';` so i'll have to ask the graph explorer maintainers if they would want to change the base url.

File: src/app/views/app-sections/AppTitle.tsx (hunk 9 above)
Comment (lhao03) [reply src/app/views/app-sections/AppTitle.tsx]: i think this was added by another member when we were trying out stuff like a toggle so i've removed it because we are no longer using the toggle. but for the future i'll remember to keep css seperate!

File: src/app/services/actions/permissions-action-creator.ts (hunk 13 above)
Comment (lhao03) [reply src/app/services/actions/permissions-action-creator.ts]: i will look into that!

File: src/app/services/graph-constants.ts (hunk 6 above)
Comment (lhao03) [reply src/app/services/graph-constants.ts]: after discussion in sprint grooming, i think ill leave in personal because its on our backlog to support delegatedPersonal perms after the MVP.

File: src/types/rsc-apps.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+export interface IAppItem extends App {
+  name: string;
+}
+
+interface App {
+  url: string;
+}
```
Comment (lhao03) [inline src/types/rsc-apps.ts]: @gracet9182 did you add this? where was/is this being used?

File: src/app/services/reducers/permission-mode-reducer.ts (hunk 8 above)
Comment (lhao03) [reply src/app/services/reducers/permission-mode-reducer.ts]: sadge LOL i will add line

File: src/types/rsc-apps.ts (hunk 14 above)
Comment (lhao03) [reply src/types/rsc-apps.ts]: i don't think its being used so i deleted it

File: src/app/views/authentication/profile/Profile.tsx (hunk 4 above)
Comment (lhao03) [reply src/app/views/authentication/profile/Profile.tsx]: i have deleted it

**Comment by lhao03** [comment]
hey @acchiang and @ChangshengYan, thank you so much for your reviews! i've changed the code accordingly, let me know if there's anything else.

**Review by csyan5 — APPROVED** [review]
In the future could we have more tests (unit tests as what this PR did, as well as some end to end tests by using automation tools like Selenium https://www.selenium.dev/)
